File: Diseño  columnas legacy/frontend/src/api/projects.ts
import { apiClient } from "./client";

export type ProjectStatus =
  | "draft"
  | "in_preparation"
  | "in_review"
  | "observed"
  | "validated"
  | "released"
  | "archived"
  | "cancelled"
  | "blocked";

export type MaturityLevel = "M0" | "M1" | "M2" | "M3" | "M4";

export interface Project {
  id: string;
  project_code: string;
  name: string;
  country: string;
  language: string;
  currency: string;
  timezone: string;
  confidentiality: string;
  status: ProjectStatus;
  maturity: MaturityLevel;
  owner_user_id: string;
  customer_id: string | null;
  opportunity_ref: string | null;
  description: string | null;
  region: string | null;
  cloned_from_id: string | null;
  created_at: string;
  updated_at: string;
  archived_at: string | null;
}

export interface PaginatedProjects {
  items: Project[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface ProjectCreateInput {
  name: string;
  country: string;
  language?: string;
  currency?: string;
  timezone?: string;
  confidentiality?: string;
  description?: string;
  region?: string;
}

export interface Revision {
  id: string;
  project_id: string;
  revision_code: string;
  revision_type: string;
  maturity: MaturityLevel;
  description: string | null;
  change_summary: string | null;
  is_frozen: boolean;
}

export interface RevisionCreateInput {
  revision_code: string;
  revision_type?: "draft" | "technical" | "client" | "production" | "as_built";
  description?: string;
  change_summary?: string;
}

export async function listProjects(): Promise<PaginatedProjects> {
  const { data } = await apiClient.get<PaginatedProjects>("/projects");
  return data;
}

export async function getProject(projectId: string): Promise<Project> {
  const { data } = await apiClient.get<Project>(`/projects/${projectId}`);
  return data;
}

export async function createProject(input: ProjectCreateInput): Promise<Project> {
  const { data } = await apiClient.post<Project>("/projects", input);
  return data;
}

export async function transitionProjectStatus(
  projectId: string,
  targetStatus: ProjectStatus,
  reason: string,
): Promise<Project> {
  const { data } = await apiClient.post<Project>(`/projects/${projectId}/status`, {
    target_status: targetStatus,
    reason,
  });
  return data;
}

export async function listRevisions(projectId: string): Promise<{ items: Revision[] }> {
  const { data } = await apiClient.get<{ items: Revision[] }>(`/projects/${projectId}/revisions`);
  return data;
}

export async function createRevision(
  projectId: string,
  input: RevisionCreateInput,
): Promise<Revision> {
  const { data } = await apiClient.post<Revision>(`/projects/${projectId}/revisions`, input);
  return data;
}

File: Diseño  columnas legacy/backend/app/models/schemas/projects.py
"""
Salvi Studio · Columns — Schemas Pydantic para proyectos
Toda magnitud incluye value + unit (P-06, sección 10 Fase 1).
"""
import uuid
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, field_validator, model_validator

from app.core.security import MaturityLevel, ProjectStatus


# ── Proyecto ────────────────────────────────────────────────────────────────

class ProjectCreate(BaseModel):
    name: str = Field(..., max_length=180)
    country: str = Field(..., min_length=2, max_length=2, description="ISO 3166-1 alpha-2")
    language: str = Field(default="es", pattern="^(es|en|fr|ca|it|pt)$")
    currency: str = Field(default="EUR", min_length=3, max_length=3)
    timezone: str = Field(default="Europe/Madrid")
    confidentiality: str = Field(default="internal", pattern="^(internal|restricted|client)$")
    description: Optional[str] = None
    customer_id: Optional[uuid.UUID] = None
    opportunity_ref: Optional[str] = Field(default=None, max_length=120)
    region: Optional[str] = Field(default=None, max_length=80)
    cloned_from_id: Optional[uuid.UUID] = None


class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=180)
    description: Optional[str] = None
    customer_id: Optional[uuid.UUID] = None
    opportunity_ref: Optional[str] = Field(default=None, max_length=120)
    confidentiality: Optional[str] = Field(default=None, pattern="^(internal|restricted|client)$")
    region: Optional[str] = Field(default=None, max_length=80)


class ProjectRead(BaseModel):
    id: uuid.UUID
    project_code: str
    name: str
    country: str
    language: str
    currency: str
    timezone: str
    confidentiality: str
    status: ProjectStatus
    maturity: MaturityLevel
    owner_user_id: uuid.UUID
    customer_id: Optional[uuid.UUID]
    opportunity_ref: Optional[str]
    description: Optional[str]
    region: Optional[str]
    cloned_from_id: Optional[uuid.UUID]
    created_at: datetime
    updated_at: datetime
    archived_at: Optional[datetime]

    model_config = {"from_attributes": True}


class ProjectStatusTransition(BaseModel):
    """Solicitud de cambio de estado con razón obligatoria."""
    target_status: ProjectStatus
    reason: str = Field(..., min_length=1, max_length=1000)


class ProjectMaturityTransition(BaseModel):
    """Solicitud de cambio de nivel de madurez."""
    target_maturity: MaturityLevel
    reason: str = Field(..., min_length=1, max_length=1000)
    validation_comment: Optional[str] = None  # Obligatorio para M3


# ── Escenario ───────────────────────────────────────────────────────────────

class ScenarioCreate(BaseModel):
    name: str = Field(..., max_length=180)
    description: Optional[str] = None
    site_id: Optional[uuid.UUID] = None
    is_base: bool = False
    cloned_from_id: Optional[uuid.UUID] = None


class ScenarioRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    description: Optional[str]
    status: str
    is_base: bool
    site_id: Optional[uuid.UUID]
    cloned_from_id: Optional[uuid.UUID]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Revisión ─────────────────────────────────────────────────────────────────

class RevisionCreate(BaseModel):
    revision_code: str = Field(..., max_length=16, description="Ej: R00, D1, C00")
    revision_type: str = Field(
        default="draft",
        pattern="^(draft|technical|client|production|as_built)$"
    )
    description: Optional[str] = None
    change_summary: Optional[str] = None


class RevisionRead(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    revision_code: str
    revision_type: str
    maturity: MaturityLevel
    description: Optional[str]
    change_summary: Optional[str]
    is_frozen: bool
    frozen_at: Optional[datetime]
    frozen_by_id: Optional[uuid.UUID]
    validated_at: Optional[datetime]
    validated_by_id: Optional[uuid.UUID]
    input_hash: Optional[str]
    created_at: datetime

    model_config = {"from_attributes": True}


class RevisionFreezeRequest(BaseModel):
    """
    Congelar una revisión — P-01.
    Una vez congelada, ninguna modificación es posible: cambios crean nueva revisión.
    """
    change_summary: str = Field(..., min_length=1, max_length=2000)
    maturity: MaturityLevel


class RevisionValidateRequest(BaseModel):
    """Validación OT — transición a M3."""
    validation_comment: str = Field(..., min_length=1, max_length=2000)
    accept: bool = True  # False = rechazar con observaciones


# ── Respuestas paginadas ─────────────────────────────────────────────────────

class PaginatedProjects(BaseModel):
    items: List[ProjectRead]
    total: int
    page: int
    page_size: int
    pages: int


class PaginatedRevisions(BaseModel):
    items: List[RevisionRead]
    total: int
    page: int
    page_size: int
    pages: int
